### PR TITLE
Multiple model selection with remote service

### PR DESCRIPTION
**Pull request by sgurunat**
Description
This PR contains changes related to multiple models selection in ProductivitySuite ChatQnA along with some minor enhancements to the UI. Also it contains docker compose files and instructions to run ProductivitySuite on Intel Gaudi server with remote TGI/TEI services.

Type of change
 New feature (non-breaking change which adds new functionality)
 Others (enhancement, documentation, validation, etc.)
New Features:

Add chatqna_wrapper.py along with updated Dockerfile.wrapper. To support multiple models chatqna with wrapper is required
ProductivitySuite: Add docker compose files for Intel Gaudi server along with remote tgi/tei service with instructions
ProductivitySuite UI: Add multiple models support. Choose different models from dropdown
Enhancements:

ProductivitySuite UI: Update names of ChatQnA, CodeGen, DocSum to Digital Assistant, Code Generator, Content Summarizer respectively
ProductivitySuite UI: Update Docsum to have vertical scroll bar if content exceeds the window height
ProductivitySuite UI: Remove <|eot_id|> string from the Chat, Docsum and Faqgen response
ProductivitySuite UI: Update contextWrapper and contextTitle width to adjust to different screen sizes
ProductivitySuite UI: Show system prompt input field always to edit in the chatqna prompt section
ProductivitySuite UI: Update max_new_tokens into max_tokens

### CODE REVIEW

File: ProductivitySuite/docker_compose/intel/hpu/gaudi/compose.yaml [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+      - tei-embedding-service
+    ports:
+      - "6000:6000"
+    ipc: host
+    environment:
+      no_proxy: ${no_proxy}
+      http_proxy: ${http_proxy}
+      https_proxy: ${https_proxy}
+      TEI_EMBEDDING_ENDPOINT: ${TEI_EMBEDDING_ENDPOINT}
+      LANGCHAIN_API_KEY: ${LANGCHAIN_API_KEY}
```
[the remark this answers is not in the document]
Comment (sgurunat) [reply ProductivitySuite/docker_compose/intel/hpu/gaudi/compose.yaml]: Thanks for the review. Removed langchain related environment variables

File: ProductivitySuite/docker_compose/intel/hpu/gaudi/set_env_remote.sh [hunk 2]
```diff
@@ -0,0 +1,57 @@
+# Copyright (C) 2024 Intel Corporation
```
Comment (letonghan) [inline ProductivitySuite/docker_compose/intel/hpu/gaudi/set_env_remote.sh:1]: Do we need different files to support this feature? I think it's better to integrate them into one file and get environment variables automatically.

Comment (sgurunat) [reply ProductivitySuite/docker_compose/intel/hpu/gaudi/set_env_remote.sh:1]: I created a separate file because in the actual file (to run all services locally) we use a variable called {host_ip} which will point to local machine host ip and different port is configured for different services. But in case of remote host there might be a different endpoints/ports for each service i.e for embedding different endpoint, for llm different endpoint etc.. Because of which I created a separate file with required environment variables so that there is no confusion for users to run it locally or remote. Please let me know if you still feel to integrate it?

Comment (letonghan) [reply ProductivitySuite/docker_compose/intel/hpu/gaudi/set_env_remote.sh:1]: Thanks for your explanation. I still think it should be integrated into one file, and explain the ${host_ip}/${remote_host_ip} stuff in the `README.md` file. Because for the remote scenario, users should be able to set any service as a remote service, not only tgi/tei embedding.

Further more, please refer to [this pr](https://github.com/opea-project/GenAIExamples/pull/977/files#diff-3c252d2e79c2d8f15044066fb23271b63448d47fc6da0a7410e976c33f40f2e0). The environment variables for docker compose have been replaced with service name, and each examples should follow this schema for users to deploy services easier.

Thank you.

**Review by ftian1 — APPROVED** [review]

File: ChatQnA/Dockerfile.wrapper [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    libjemalloc-dev \
+    git
+
+RUN useradd -m -s /bin/bash user && \
+    mkdir -p /home/user && \
+    chown -R user /home/user/
+
+COPY /GenAIComps /home/user/GenAIComps
+WORKDIR /home/user/
+#RUN git clone --branch v1.0 https://github.com/opea-project/GenAIComps.git
```
Comment (amberjain1) [inline ChatQnA/Dockerfile.wrapper:19]: We shall get it from clone itself. This line shall not be commented out.

File: ChatQnA/Dockerfile.wrapper [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+RUN apt-get update -y && apt-get install -y --no-install-recommends --fix-missing \
+    libgl1-mesa-glx \
+    libjemalloc-dev \
+    git
+
+RUN useradd -m -s /bin/bash user && \
+    mkdir -p /home/user && \
+    chown -R user /home/user/
+
+COPY /GenAIComps /home/user/GenAIComps
```
Comment (amberjain1) [inline ChatQnA/Dockerfile.wrapper:17]: Dont  copy. Instead clone it.

File: ProductivitySuite/docker_compose/intel/hpu/gaudi/README.md [hunk 5]
```diff
@@ -0,0 +1,554 @@
...
+
+### 3. Build Rerank Image
+
+```bash
+docker build --no-cache -t opea/reranking-tei:latest --build-arg https_proxy=$https_proxy --build-arg http_proxy=$http_proxy -f comps/reranks/tei/Dockerfile .
+```
+
+### 4. Build LLM Image
+
+#### Use TGI as backend
```
Comment (amberjain1) [inline ProductivitySuite/docker_compose/intel/hpu/gaudi/README.md:33]: Our main example is based on VLLM. Lets add VLLM as default step and TGI as demonstration that others shall work in a similar way.

File: ProductivitySuite/docker_compose/intel/hpu/gaudi/README.md [hunk 6]
```diff
@@ -0,0 +1,554 @@
...
+```bash
+cd GenAIExamples/FaqGen
+docker build --no-cache -t opea/faqgen:latest --build-arg https_proxy=$https_proxy --build-arg http_proxy=$http_proxy -f Dockerfile .
+```
+
+### 9. Build UI Docker Image
+
+Build frontend Docker image that enables via below command:
+
+**Export the value of the public IP address of your server to the `host_ip` environment variable**
```
Comment (amberjain1) [inline ProductivitySuite/docker_compose/intel/hpu/gaudi/README.md:95]: Lets enhance the documentation for both remote and local model deployment.

File: ProductivitySuite/docker_compose/intel/hpu/gaudi/compose_remote.yaml [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+      https_proxy: ${https_proxy}
+      TEI_RERANKING_ENDPOINT: ${TEI_RERANKING_ENDPOINT}
+      HUGGINGFACEHUB_API_TOKEN: ${HUGGINGFACEHUB_API_TOKEN}
+      HF_HUB_DISABLE_PROGRESS_BARS: 1
+      HF_HUB_ENABLE_HF_TRANSFER: 0
+      CLIENTID: ${CLIENTID}
+      CLIENT_SECRET: ${CLIENT_SECRET}
+      TOKEN_URL: ${TOKEN_URL}
+    restart: unless-stopped
+  llm:
```
Comment (amberjain1) [inline ProductivitySuite/docker_compose/intel/hpu/gaudi/compose_remote.yaml:78]: should this be changed to vllm

File: ProductivitySuite/docker_compose/intel/hpu/gaudi/compose_remote.yaml [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+      http_proxy: ${http_proxy}
+      MEGA_SERVICE_HOST_IP: ${MEGA_SERVICE_HOST_IP}
+      EMBEDDING_SERVICE_HOST_IP: ${EMBEDDING_SERVICE_HOST_IP}
+      RETRIEVER_SERVICE_HOST_IP: ${RETRIEVER_SERVICE_HOST_IP}
+      RERANK_SERVICE_HOST_IP: ${RERANK_SERVICE_HOST_IP}
+      LLM_SERVICE_HOST_IP: ${LLM_SERVICE_HOST_IP_CHATQNA}
+    ipc: host
+    restart: always
+  llm_codegen:
+    image: ${REGISTRY:-opea}/llm-tgi:${TAG:-latest}
```
Comment (amberjain1) [inline ProductivitySuite/docker_compose/intel/hpu/gaudi/compose_remote.yaml:121]: change it to be vllm based

File: ProductivitySuite/docker_compose/intel/hpu/gaudi/compose_tgi_remote.yaml
Comment (amberjain1) [inline ProductivitySuite/docker_compose/intel/hpu/gaudi/compose_tgi_remote.yaml:1]: whats the difference between compose_remote and compose_remote_tgi?

File: ChatQnA/Dockerfile.wrapper (hunk 3 above)
Comment (sgurunat) [reply ChatQnA/Dockerfile.wrapper]: Uncommented the git clone command

File: ChatQnA/Dockerfile.wrapper (hunk 4 above)
Comment (sgurunat) [reply ChatQnA/Dockerfile.wrapper]: Removed the COPY command line

File: ProductivitySuite/docker_compose/intel/hpu/gaudi/compose_remote.yaml (hunk 7 above)
Comment (sgurunat) [reply ProductivitySuite/docker_compose/intel/hpu/gaudi/compose_remote.yaml:84]: Yes. Updated it to vllm. Now compose_remote.yaml contains vllm related instructions

File: ProductivitySuite/docker_compose/intel/hpu/gaudi/compose_remote.yaml (hunk 8 above)
Comment (sgurunat) [reply ProductivitySuite/docker_compose/intel/hpu/gaudi/compose_remote.yaml]: Updated it to have vllm

File: ProductivitySuite/docker_compose/intel/hpu/gaudi/compose_tgi_remote.yaml
Comment (sgurunat) [reply ProductivitySuite/docker_compose/intel/hpu/gaudi/compose_tgi_remote.yaml:1]: Deleted this file

File: ProductivitySuite/docker_compose/intel/hpu/gaudi/README.md (hunk 6 above)
Comment (sgurunat) [reply ProductivitySuite/docker_compose/intel/hpu/gaudi/README.md:107]: Updated the README.md file

File: ProductivitySuite/docker_compose/intel/hpu/gaudi/README.md (hunk 5 above)
Comment (sgurunat) [reply ProductivitySuite/docker_compose/intel/hpu/gaudi/README.md]: Updated the README.md file

File: ProductivitySuite/docker_compose/intel/hpu/gaudi/set_env_remote.sh (hunk 2 above)
Comment (sgurunat) [reply ProductivitySuite/docker_compose/intel/hpu/gaudi/set_env_remote.sh]: Removed set_env_remote.sh and handled it in the README.md file itself.

**Review by lvliang-intel — APPROVED** [review]

**Comment by lvliang-intel** [comment]
@sgurunat,
Please fix this path check issue.
image

**Comment by sgurunat** [comment]
@lvliang-intel - Fixed it

**Comment by chensuyue** [comment]
Please check the failed CI, https://github.com/opea-project/GenAIExamples/actions/runs/11795779662/job/32856256339?pr=1049

File: ProductivitySuite/docker_compose/intel/hpu/gaudi/compose.yaml [hunk 9]
```diff
@@ -0,0 +1,376 @@
+# Copyright (C) 2024 Intel Corporation
```
Comment (chensuyue) [inline ProductivitySuite/docker_compose/intel/hpu/gaudi/compose.yaml:1]: Please add a `test_compose_on_gaudi.sh` to cover the new code.

Comment (chensuyue) [reply ProductivitySuite/docker_compose/intel/hpu/gaudi/compose.yaml:1]: like this one: https://github.com/opea-project/GenAIExamples/blob/53500d108d3075b7ced0faff8433095398480b2a/ProductivitySuite/tests/test_compose_on_xeon.sh

File: ChatQnA/Dockerfile.wrapper [hunk 10]
```diff
@@ -0,0 +1,34 @@
+
```
Comment (chensuyue) [inline ChatQnA/Dockerfile.wrapper:1]: Why did you need this one? We default use no wrapper mode for current examples.

Comment (sgurunat) [reply ChatQnA/Dockerfile.wrapper:1]: For remote TGI or vLLM inference endpoints if there is oAuth authentication flow we need to get the token and pass it in the headers. Currently this functionality is added in wrappers.

Comment (chensuyue) [reply ChatQnA/Dockerfile.wrapper:1]: Did you use docker image build by this dockerfile in this PR?

Comment (chensuyue) [reply ChatQnA/Dockerfile.wrapper:1]: If yes, you should add into the docker_image_build/build.yaml, and give it a new image name.

File: ProductivitySuite/docker_compose/intel/hpu/gaudi/compose.yaml (hunk 9 above)
Comment (sgurunat) [reply ProductivitySuite/docker_compose/intel/hpu/gaudi/compose.yaml:1]: Added it

File: ChatQnA/Dockerfile.wrapper (hunk 10 above)
Comment (sgurunat) [reply ChatQnA/Dockerfile.wrapper:1]: yes added into the build.yaml file

File: ProductivitySuite/tests/test_compose_on_gaudi.sh [hunk 11]
```diff
@@ -0,0 +1,391 @@
...
+WORKPATH=$(dirname "$PWD")
+LOG_PATH="$WORKPATH/tests"
+ip_address=$(hostname -I | awk '{print $1}')
+
+function build_docker_images() {
+    cd $WORKPATH/docker_image_build
+    git clone https://github.com/opea-project/GenAIComps.git && cd GenAIComps && git checkout "${opea_branch:-"main"}" && cd ../
+
+    echo "Build all the images with --no-cache, check docker_image_build.log for details..."
+    docker compose -f build_vllm.yaml build --no-cache > ${LOG_PATH}/docker_image_build.log
```
Comment (chensuyue) [inline ProductivitySuite/tests/test_compose_on_gaudi.sh:22]: `build_vllm.yaml` -> `build.yaml` and just build the images required by this test.

File: ProductivitySuite/tests/test_compose_on_gaudi.sh [hunk 12]
```diff
@@ -0,0 +1,391 @@
...
+
+    docker pull ghcr.io/huggingface/tei-gaudi:latest
+    docker pull opea/vllm-hpu:latest
```
Comment (chensuyue) [inline ProductivitySuite/tests/test_compose_on_gaudi.sh:25]: You don't need pull this, this will be build in the CI test.

File: ProductivitySuite/tests/test_compose_on_gaudi.sh (hunk 11 above)
Comment (sgurunat) [reply ProductivitySuite/tests/test_compose_on_gaudi.sh]: Oh yeah, missed to change it. Updated it now. Thanks

File: ProductivitySuite/tests/test_compose_on_gaudi.sh (hunk 12 above)
Comment (sgurunat) [reply ProductivitySuite/tests/test_compose_on_gaudi.sh]: ok commented it out.

**Comment by chensuyue** [comment]
This PR should be closed right?
